### PR TITLE
arm64/a64: Add driver for MIPI DSI

### PR DESCRIPTION
**Pull request by lupyuen**
## Summary

This PR adds the driver for Allwinner A64's MIPI Display Serial Interface (DSI) and MIPI Display Physical Layer (D-PHY).

This driver will be used by the upcoming Display Driver for PINE64 PinePhone.

### Modified Files

- `arch/arm64/src/a64/Kconfig`: Added the Kconfig option for "A64 Peripheral Selection > MIPI DSI" (`CONFIG_A64_MIPI_DSI`), which enables the MIPI DSI Driver

- `arch/arm64/src/a64/hardware/a64_memorymap.h`: Added the Base Address for MIPI DSI

- `arch/arm64/src/a64/Make.defs`: Added the MIPI DSI Driver to the Makefile

### New Files

- `arch/arm64/src/a64/mipi_dsi.c`, `mipi_dsi.h`: Compose MIPI DSI Packets (Long, Short, Short with Parameter)

- `arch/arm64/src/a64/a64_mipi_dsi.c`, `a64_mipi_dsi.h`: MIPI DSI Driver for Allwinner A64

- `arch/arm64/src/a64/a64_mipi_dphy.c`, `a64_mipi_dphy.h`: MIPI D-PHY Driver for Allwinner A64

### Updated Documentation

- `platforms/arm/a64/boards/pinephone/index.rst`: Added MIPI DSI as supported peripheral for PinePhone

## Impact

Our MIPI DSI Driver will be built only when we select the Kconfig option for "A64 Peripheral Selection > MIPI DSI" (`CONFIG_A64_MIPI_DSI`).

The MIPI DSI Driver will be called by the upcoming drivers for A64 Timing Controller (TCON) and A64 Display Engine.

Our MIPI DSI Driver has no impact on existing code because it's not called yet by the existing code.

The Driver Code is derived from my Reverse-Engineering of the PinePhone Display Driver, based on the logs captured from the PinePhone p-boot Bootloader:

-   ["Understanding PinePhone's Display (MIPI DSI)"](https://lupyuen.github.io/articles/dsi)

-   ["NuttX RTOS for PinePhone: Display Driver in Zig"](https://lupyuen.github.io/articles/dsi2)

## Testing

Some parts of the PinePhone Display Driver are still under construction. To test the MIPI DSI Driver, we ran a Zig Program that integrates the NuttX MIPI DSI Driver with the upcoming parts of the Display Driver:

-  ["Test MIPI DSI for NuttX Kernel"](https://github.com/lupyuen/pinephone-nuttx#test-mipi-dsi-for-nuttx-kernel)

The Integration Test [successfully rendered graphics](https://lupyuen.github.io/images/dsi3-title.jpg) on the LCD Display. Here's the Test Log:

-  [Test Log for NuttX MIPI DSI on PinePhone](https://gist.github.com/lupyuen/f1a02068aeb0785278c482116a4eedc7)

We also ran Unit Tests for the MIPI DSI Driver, [as explained here.](https://github.com/lupyuen/pinephone-nuttx#test-mipi-dsi-for-nuttx-kernel)

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@lupyuen please rebase to the last mater to fix macOS ci break.

**Comment by lupyuen** [comment]
@xiaoxiang781216 Thanks, hope I did this correctly

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 1]
```diff
@@ -0,0 +1,984 @@
...
+/* DSI Line Number Register 1 (A31 Page 847) */
+#define DSI_BASIC_SIZE1_REG (A64_DSI_ADDR + 0x1c)
+#define VIDEO_VACT(n)       (n << 0)
+#define VIDEO_VT(n)         (n << 16)
+
+/* DSI Instruction Function Register (Undocumented) */
+#define DSI_INST_FUNC_REG(n)   (A64_DSI_ADDR + 0x020 + n * 0x04)
+#define DSI_INST_FUNC_LANE_CEN (1 << 4)
+
+/* DSI Instruction Loop Select Register (Undocumented) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:107]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 2]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_INST_LOOP_SEL_REG  (A64_DSI_ADDR + 0x40)
+
+/* DSI Instruction Loop Number Register (Undocumented) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:110]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 3]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_INST_LOOP_NUM_REG(n) (A64_DSI_ADDR + 0x44 + n * 0x10)
+
+/* DSI Instruction Jump Select Register (Undocumented) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:113]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 4]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_INST_ID_TBA       1
+#define DSI_INST_ID_HSC       2
+#define DSI_INST_ID_HSD       3
+#define DSI_INST_ID_LPDT      4
+#define DSI_INST_ID_HSCEXIT   5
+#define DSI_INST_ID_NOP       6
+#define DSI_INST_ID_DLY       7
+#define DSI_INST_ID_END       15
+
+/* DSI Instruction Jump Configuration Register (Undocumented) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:125]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 5]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_INST_JUMP_CFG_REG(n) (A64_DSI_ADDR + 0x4c + n * 0x04)
+#define DSI_INST_JUMP_CFG        0
+
+/* DSI Transfer Start Register (Undocumented) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:129]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 6]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_TRANS_START_REG (A64_DSI_ADDR + 0x60)
+
+/* DSI Transfer Zero Register (Undocumented) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:132]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 7]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_TRANS_ZERO_REG  (A64_DSI_ADDR + 0x78)
+
+/* DSI Timing Controller DRQ Register (Undocumented) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:135]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 8]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_TCON_DRQ_REG    (A64_DSI_ADDR + 0x7c)
+
+/* DSI Pixel Format Register 0 (A31 Page 847) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:138]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 9]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_PIXEL_CTL0_REG (A64_DSI_ADDR + 0x80)
+#define PIXEL_FORMAT(n) (n << 0)
+#define PIXEL_ENDIAN    (0 << 4)
+#define PD_PLUG_DIS     (1 << 16)
+
+/* DSI Pixel Package Register 0 (A31 Page 848) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:144]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 10]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_PIXEL_PH_REG (A64_DSI_ADDR + 0x90)
+#define PIXEL_DT(n)      (n << 0)
+#define PIXEL_VC(n)      (n << 6)
+#define PIXEL_WC(n)      (n << 8)
+#define PIXEL_ECC(n)     (n << 24)
+
+/* DSI Pixel Package Register 2 (A31 Page 849) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:151]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 11]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_PIXEL_PF0_REG (A64_DSI_ADDR + 0x98)
+#define CRC_FORCE         0xffff
+
+/* DSI Pixel Package Register 3 (A31 Page 849) */
+#define DSI_PIXEL_PF1_REG (A64_DSI_ADDR + 0x9c)
+#define CRC_INIT_LINE0(n) (n << 0)
+#define CRC_INIT_LINEN(n) (n << 16)
+
+/* DSI Sync Package Register 0 (A31 Page 850) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:160]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 12]
```diff
@@ -0,0 +1,984 @@
...
+#define PIXEL_DT(n)      (n << 0)
+#define PIXEL_VC(n)      (n << 6)
+#define PIXEL_WC(n)      (n << 8)
+#define PIXEL_ECC(n)     (n << 24)
+
+/* DSI Pixel Package Register 2 (A31 Page 849) */
+#define DSI_PIXEL_PF0_REG (A64_DSI_ADDR + 0x98)
+#define CRC_FORCE         0xffff
+
+/* DSI Pixel Package Register 3 (A31 Page 849) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:155]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 13]
```diff
@@ -0,0 +1,984 @@
...
+
+/* DSI Sync Package Register 0 (A31 Page 850) */
+#define DSI_SYNC_HSS_REG (A64_DSI_ADDR + 0xb0)
+#define SYNC_ECC(n)      (n << 24)
+#define SYNC_D1(n)       (n << 16)
+#define SYNC_D0(n)       (n << 8)
+#define SYNC_VC(n)       (n << 6)
+#define SYNC_DT(n)       (n << 0)
+
+/* DSI Sync Package Register 1 (A31 Page 850) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:168]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 14]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_SYNC_HSE_REG (A64_DSI_ADDR + 0xb4)
+
+/* DSI Sync Package Register 2 (A31 Page 851) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:171]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 15]
```diff
@@ -0,0 +1,984 @@
...
+#define DSI_SYNC_VSS_REG (A64_DSI_ADDR + 0xb8)
+
+/* DSI Sync Package Register 3 (A31 Page 851) */
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:174]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 16]
```diff
@@ -0,0 +1,984 @@
...
+static void a64_disable_dsi_processing(void)
+{
+  /* DSI Configuration Register 0 (A31 Page 845)
+   * Set INSTRU_EN (Bit 0) to 0
+   * (Disable DSI Processing)
+   */
+
+  modreg32(0,
+           INSTRU_EN,
+           DSI_BASIC_CTL0_REG);
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:257]: This three lines could fit in a single line lesser than 78 chars

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 17]
```diff
@@ -0,0 +1,984 @@
...
+static void a64_enable_dsi_processing(void)
+{
+  /* DSI Configuration Register 0 (A31 Page 845)
+   * Set INSTRU_EN (Bit 0) to 1
+   * (Enable DSI Processing from Instruction 0)
+   */
+
+  modreg32(INSTRU_EN,
+           INSTRU_EN,
+           DSI_BASIC_CTL0_REG);
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:283]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 18]
```diff
@@ -0,0 +1,984 @@
...
+                        (b[1] << 8) +
+                        (b[2] << 16) +
+                        (b[3] << 24);
+
+      /* Write the 32-bit value */
+
+      DEBUGASSERT(addr <= DSI_CMD_TX_END);
+      modreg32(v,
+               0xffffffff,
+               addr);
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:461]: Ditto

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 19]
```diff
@@ -0,0 +1,984 @@
...
+      addr += 4;
+    }
+
+  /* DSI Low Power Control Register (A31 Page 854)
+   * Set TX_Size (Bits 0 to 7) to Packet Length - 1
+  */
+
+  modreg32(pktlen - 1,
+           0xff,
+           DSI_CMD_CTL_REG);
```
Comment (acassis) [inline arch/arm64/src/a64/a64_mipi_dsi.c:471]: Ditto

**Comment by xiaoxiang781216** [comment]
Please rebase to the last mater again to fix ci broken.

**Comment by xiaoxiang781216** [comment]
@lupyuen could you squash all change into one commit?

**Comment by lupyuen** [comment]
@xiaoxiang781216 GitHub Desktop won't let me squash the commits after merging. But no worries I'll copy the files and submit a new PR. Thanks :-)

**Pull request closed without merging**